tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc

```diff
@@ -81,23 +136,53 @@ __global__ void SortedSegmentSumCustomKernel(
             last_output_segment_id * inner_dim_size + segment_offset;
         // decide whether to write result to global memory using atomic
         // operations
-        if (last_output_segment_id == first_segment_id) {
-          GpuAtomicAdd(output + output_index, sum);
+        if (last_output_segment_id == first_segment_id 
+            && is_first_update_racy) {
+          Index previous_segment_num = 0;
+          for (Index k = 1; input_outer_dim_index_base - k >= 0; k++) {
+            if (segment_ids[input_outer_dim_index_base]
+                  != segment_ids[input_outer_dim_index_base - k])
+              break;
+            previous_segment_num++;
+          }
+          cur_segment_num += previous_segment_num;
+          GpuAtomicAdd(output + output_index, sum / T(cur_segment_num));
```

Comment:
This looks fishy to me -- you can't increment a mean by another mean.  If there is some trick here, can you please add a comment making that clear?